modules/codegen/src/main/scala/com/twilio/guardrail/generators/Java/AsyncHttpClientClientGenerator.scala

```diff
@@ -0,0 +1,770 @@
+import com.github.javaparser.ast.stmt._
+import com.twilio.guardrail.SwaggerUtil.jpaths
+import com.twilio.guardrail.generators.{Response, ScalaParameter}
+import com.twilio.guardrail.generators.syntax.Java._
+import com.twilio.guardrail.languages.JavaLanguage
+import com.twilio.guardrail.protocol.terms.client._
+import com.twilio.guardrail.terms.RouteMeta
+import com.twilio.guardrail.{RenderedClientOperation, StaticDefns, SupportDefinition, Target}
+import java.net.URI
+import java.util
+
+object AsyncHttpClientClientGenerator {
+  private val URI_TYPE = JavaParser.parseClassOrInterfaceType("URI")
+  private val DEFAULT_ASYNC_HTTP_CLIENT_CONFIG_BUILDER_TYPE = JavaParser.parseClassOrInterfaceType("DefaultAsyncHttpClientConfig.Builder")
+  private val DEFAULT_ASYNC_HTTP_CLIENT_TYPE = JavaParser.parseClassOrInterfaceType("DefaultAsyncHttpClient")
+  private val ASYNC_HTTP_CLIENT_TYPE = JavaParser.parseClassOrInterfaceType("AsyncHttpClient")
+  private val ASYNC_HTTP_CLIENT_CONFIG_TYPE = JavaParser.parseClassOrInterfaceType("AsyncHttpClientConfig")
+  private val REQUEST_BUILDER_TYPE = JavaParser.parseClassOrInterfaceType("RequestBuilder")
+  private val REQUEST_TYPE = JavaParser.parseClassOrInterfaceType("Request")
+  private val RESPONSE_TYPE = JavaParser.parseClassOrInterfaceType("Response")
+  private val FILE_PART_TYPE = JavaParser.parseClassOrInterfaceType("FilePart")
+  private val STRING_PART_TYPE = JavaParser.parseClassOrInterfaceType("StringPart")
+  private val OBJECT_MAPPER_TYPE = JavaParser.parseClassOrInterfaceType("ObjectMapper")
+  private val BUILDER_TYPE = JavaParser.parseClassOrInterfaceType("Builder")
+  private val MARSHALLING_EXCEPTION_TYPE = JavaParser.parseClassOrInterfaceType("MarshallingException")
+  private val HTTP_ERROR_TYPE = JavaParser.parseClassOrInterfaceType("HttpError")
+  private val EXCEPTION_TYPE = JavaParser.parseClassOrInterfaceType("Exception")
+  private val ILLEGAL_ARGUMENT_EXCEPTION_TYPE = JavaParser.parseClassOrInterfaceType("IllegalArgumentException")
+  private val JSON_PROCESSING_EXCEPTION_TYPE = JavaParser.parseClassOrInterfaceType("JsonProcessingException")
+
+  private val HTTP_CLIENT_FUNCTION_TYPE = functionType.setTypeArguments(new NodeList[Type](
+      REQUEST_TYPE,
+      completionStageType.setTypeArguments(RESPONSE_TYPE)
+    ))
+
+  private def typeReferenceType(typeArg: Type): ClassOrInterfaceType =
+    JavaParser.parseClassOrInterfaceType("TypeReference").setTypeArguments(typeArg)
+
+  private def showParam(param: ScalaParameter[JavaLanguage], overrideParamName: Option[String] = None): Expression = {
+    val paramName = overrideParamName.getOrElse(param.paramName.asString)
+    new MethodCallExpr(
+      new MethodCallExpr(new NameExpr("Shower"), "getInstance"),
+      "show",
+      new NodeList[Expression](new NameExpr(paramName))
+    )
+
+    def doShow(tpe: Type): Expression = tpe match {
+      case cls: ClassOrInterfaceType if cls.isOptional || cls.isNamed("List") =>
```

Comment:
It would, unfortunately.  Should try to qualify it I guess?

Reply:
Yeah -- an additional question is that we don't support showables for implementers of an interface, preventing `java.util.List` from actually being registered as showable, fwict

Reply:
There's nothing to prevent anyone from registering any concrete (or non-concrete) class as a `Showable`, but yeah, `List` is handled in a special way with behavior a user couldn't add.  Unfortunately javaparser has no idea what a List is (or, better, what an Iterable is), so I'm not sure what to do here.

Reply:
This means any parameterized type will not be able to integrate, right? If `Showable<A>` had both `String show(A o)` and `Boolean canShow(A o)`, I could

```scala
shower.register(classOf[List[_]], new Showable[List[_]] {
    def canShow(l: List[_]): Boolean = l.headOption.fold(true)(o => shower.canShow(o.getClass))
    def show(l: List[_]): String = s"List(${l.map(shower.show(_)).mkString(", ")})"
)
```

similar to how `unapply` works

Reply:
But that doesn't actually do anything useful here, at least for the specific case of a list of something.  If you look at the generated code for AsyncHttpClient, the code that gets emitted in the special case of `List` is (roughly):
```
for (final ${containedType} item : list) {
    builder.addFormParam("${paramName}", Shower.show(item));
}
```
Being able to take a list and generate a string that concats the values with commas in between isn't useful.

Parameterized types that aren't iterable should work, assuming the contained type also has a `Showable` instance registered.  But again, the string that results has to actually be useful as a value.  So you could perhaps do somthing like:
```
shower.register(classOf[Option[_]], new Showable[Option[_]] {
  def show(o: Option[_]): String = o.map(shower.show).orNull
}
```
That would actually work properly for the specific case of adding a form or query param for the AHC client generator, I think.

Back to list types: what we really need is a way to identify if a (custom) type is iterable or not, and then handle that iterable in whatever way makes sense for that purpose (add each element as a form or query param, or whatever).  I'm not really sure how to do that, though, since JavaParser, when given an arbitrary string for a type, can't tell if the type implements `Iterable`.  Perhaps if we know the OpenAPI type is `array`, then we just assume that if something is passed via `x-scala-type` for it, it must be `Iterable`?  Not sure...

Reply:
At least if that assumption is wrong, it would be a discoverable type error. I'm thinking of attempts at refinement types, like `NonEmptyList` registered with Jackson such that more requirements can be expressed in the protocol, leading way to the OpenAPI validation stuff